### ISSUE TITLE
Memory problem or leak on iOS 7 ?

### ISSUE THREAD

**Issue by kefon** [rs/SDWebImage#538]
Hello everybody,
I'm sorry in advance as maybe this is a dumb and noob question...

I'm using SDWebImage to display pictures in a UITableView in my <b>cellForRowAtIndexPath</b> method, using the classic

<pre><code>[cell.pointPicture setImageWithURL:[NSURL URLWithString:thePoint.imageURL] placeholderImage:[UIImage imageNamed:POINT_DEFAULT_IMAGE]];</code></pre>

(the displayed pictures are light and well compressed jpgs, just some ko)
When I inspect my app with "<b>Instrument - Allocations</b>", and just scroll down my UITableView (with 40 cells containing picture, a bit like Instagram), I got a huge amount of memory used ! (see screenshot)

![screen](https://f.cloud.github.com/assets/5464064/1415786/c4fc3d50-3f1e-11e3-8a6d-a7e1b8d5f549.jpg)

But it seems to be "<b>VM</b>", and especially "<b>VM: CG raster data</b>" from the <b>coreGraphics</b> library.

So the questions are : 
- Is it normal?
- Is that a serious problem?
- Is there a way to avoid this?

I'm sorry but after few search on the web I can't find any relevant information concerning the "<b>VM: CG raster data</b>"... Any idea? Thanks in advance !

**Comment by delannoyk** [comment]
Had the same issue in my app (which use a lot of large JPEG images).

Thanks to @harishkashyap for the patch, works well!

**Comment by rs** [comment]
@harishkashyap could you please provide a PR?

**Comment by harishkashyap** [comment]
@rs I have put in a PR for this: https://github.com/rs/SDWebImage/pull/996. Note that this also needs the configuration to be set:
    [SDImageCache sharedImageCache].shouldDecompressImages = NO;
    [SDWebImageDownloader sharedDownloader].shouldDecompressImages = NO;

I would favor defaulting to these options in the library but that is your decision. I added these as options externally so not to have breaking changes.

**Comment by donholly** [comment]
I'm a little confused...

The entire point of the decompression step when serving images via SDWebImage is to avoid that expense when the image is drawn to screen (and to keep it off of the main thread).

You can read more about this here (the TLDR is that decompressing JPEGs is expensive):
http://www.cocoanetics.com/2011/10/avoiding-image-decompression-sickness/

What you're likely going to see when you disable this functionality is that images will take longer to appear (and may cause main thread hiccups) when you're drawing them (this will be more noticeable in a UICollectionView with many images). 

Per my understanding, this won't actually prevent CGRasterData accumulation (since this same process will occur when you draw the images to screen). At best, you're just delaying the decompression costs to the point of drawing - which has its merits, but doesn't solve the underlying memory issue in the long run. The only reason I could think you might not want to decompress the images on a background thread upon retrieving them from the cache is if you're not planning on drawing them immediately (pre-fetching for later use).

UIImage has its own memory management techniques and should purge unused image data when it is no longer needed (or when memory warnings occur). SDWebImage also uses an NSCache to store in-memory images, which will also purge its contents in the event of a memory warning.

I do *not* think image decompression should be off by default. This is one of the key benefits of using SDWebImage in the first place (decompressing JPEGs in a background thread is noticeably more performant than doing so JIT on the main thread).

I would also argue that having two separate places for setting that option (both in the cache and image downloader) is poor design and not very intuitive. Perhaps moving this option to the SDWebImageManager (which can then set it in both places on the user's behalf) is a better way to do this?

**Comment by harishkashyap** [comment]
It does prevent CGRasterData accumulation as the amount of memory persisted during infinite scrolling only increases and infact increases tremendously to a point that causes the crash. @foundry has demonstrated it and that is exactly what we found as well. Also we didn't see any stutter while testing; on the contrary it remarkably improves. Also patch allows choosing the queue and it was using the main queue to prefetch that was adding a bad performance to the already existing high memory due to decompression. The idea of having different options is to allow granular level of choice that was consciously done not a poor design. This reduced memory by almost as much as 2/3rd. The earlier implementation didn't perhaps consider inifinite scrolling with large images and that was a gross error.

**Comment by rromanchuk** [comment]
@donholly i can objectively say i have found no performance degradation turning off decompression using @harishkashyap patch. Not using the patch causes my app to crash in under one minute as the application reaches it's memory pressure limits. You can reproduce this very easily, just use larger images. `UIImage` does *not* clean up its  CGRasterData

To be honest, the article you posted may be completely invalid in today's market. I have ran many different UICollectionViews with decompression off and have noticed zero difference. The current state of `SDWebImage` using decompression with images any larger than 100x100 is going to cause ios system shutdowns as RasterData balloons to device limitations.

I honestly don't think you will see any performance degradation or FPS frame drops using instruments when turning off decompression. Without this patch we need to put a huge warning that SDWebImage will only not crash for users using small thumbnails (but will still leak unless the client listens for memory warnings and manually wipes sdwebimage)

**Comment by donholly** [comment]
Okay - perhaps I'm just not seeing this b/c of the size of the images I'm loading. What size images do you guys see these problems with? And you don't notice *any* difference in drawing time when you turn decompression off? Even with older devices (iPhone 4/4S)?

You may be right that the article I referenced is out of date. If so, I may actually want to use this patch and save CPU cycles for decompression.

**Comment by harishkashyap** [comment]
I would be happy to move the configs to as sdwebimagemanager and default to not decompressing if everyone concurs.

**Comment by delannoyk** [comment]
@donholly I had the memory issue causing crashes on iPad 2. I didn't notice any performance issue applying @harishkashyap 's patch as an user of the app.

Images I'm loading are quite large (some of them are bigger than 1000 x 1000).

**Comment by donholly** [comment]
I didn't mean to criticize the design, sorry :)

What was the reason for separating them? Why would you want them to be
decompressed from the cache but not when downloaded (or vice versa)?

**Comment by donholly** [comment]
Got it... is the consensus that background image decompression is a thing
of the past now? Perhaps ImageIO/CoreGraphics are just better/smarter about
this now?

**Comment by harishkashyap** [comment]
The other way around, to have compressed in cache and decompressed in rendering just to control cache. But I agree that is minor and will be happy to move it to the webmanager

**Comment by donholly** [comment]
Got it. Sounds good!

**Comment by holetse** [comment]
@donholly The big issue here seems to be the NSCache has a memory leak when its used to cache CGRasterData. All these solutions are just ways of working around that central bug.

**Comment by harishkashyap** [comment]
I think decompression methods may have a leak; not sure if it is NSCache. There are alternatives to + (UIImage *)decodedImageWithImage:(UIImage *)image on iOS 7+. Someone can see if the new decompression methods resolves that. However, infinite scrolling with large images + caching on decompressed images (no matter what method you use) can consume huge memory at any given time that can affect performance.

**Comment by donholly** [comment]
@harishkashyap - What alternative methods are you referring to?

**Comment by harishkashyap** [comment]
http://www.objc.io/issue-5/iOS7-hidden-gems-and-workarounds.html has decompression methods that I haven't had time to investigate as an alternative method to what is currently used.

**Comment by donholly** [comment]
Ah yes, I do remember coming across that awhile back now.. I didn't investigate further due to what @steipete noted:

```I was very excited when I first found out about this, but you really shouldn’t be. In my tests, performance actually decreased when I enabled immediate caching. Either this method calls up to the main thread (unlikely) or perceived performance is simply worse because it locks in copyImageBlockSetJPEG, which is also used when showing a non-decrypted image on the main thread.```

**Comment by harishkashyap** [comment]
Thanks @donholly! That was the only thing I was going to look at.

**Comment by bbnnt** [comment]
all those comment shows that the problem is obvious
Read it all, no real workarounds… should another pod be recommended ?

**Comment by mallorypaine** [comment]
NSCache is definitely the problem. I added a PR which eliminates the memory spikes by allowing the in-memory cache to be disabled.

**Comment by rs** [comment]
would implementing the mem cache without nscache fix the issue in you opinion?

**Comment by bbnnt** [comment]
@rs sorry to ask as I'm not that experienced in using sdwebimage;
but what's not obvious here to me, is wether there's a fix to apply to sdwebimage to play well with newer iOS version, of if this whole discussion is just a long list of people using bad patterns ?

**Comment by rs** [comment]
My understanding is that most people here see a leak where it is just cached images. Memory caching of images used to be a big deal to get 60fps on old hardware. It's certainly less the case today, and maybe disk only caching can be enough for some use cases.

**Comment by harishkashyap** [comment]
@rs that's right. This reminds me. The default should be changed to put off decompression so there is no patch that needs to be applied.

**Comment by bbnnt** [comment]
@harishkashyap thing is with this tons of posts here; I don't want to bother but I'm lost…

what would be the simplest config to apply to get this to work properly, the "put off decompression" statement ?

**Comment by rs** [comment]
@harishkashyap can you send a pull request? @mallorypaine if you disable decompression, does it fix your issue without disabling memory caching?

**Comment by foundry** [comment]
@bbnnt: this is how to disable decompression -
change: 
`_shouldDecompressImages = YES`
 to: 
`_shouldDecompressImages = NO` 
in: 
SDImageCache.m
SDWebImageDownloader.m
SDWebImageDownloaderOperation.m

This assumes use of current SDWebImage which has @harishkashyap fixes included. In my case, this fixes the issue. Without this fix, ballooning memory that resulted in crashes every time (context is infinite scrolling tableview/collectionview cells). See my post [further up the page](https://github.com/rs/SDWebImage/issues/538#issuecomment-67951641)

**Comment by mallorypaine** [comment]
@rs Yes, implementing an in-memory cache with a cost feature that is actually never exceeded would also fix this issue. For my use case, I *do* want my images forcibly decoded on a background thread. That way, I don't get scrolling stutters when the images are ultimately displayed in a UIImage View.

If I turn off decompression (which I need), then the spikes appear to go away. However, the decompression is still unrelated to this issue. As I'm sure you know, a UIImage's memory consumption balloons the first time that image is actually drawn as that's when decompression + bitmap buffer allocation takes place. If I turn off decompression, then the UIImages that are in the NSCache are just lightweight pointers referencing file data, so the spike does appear to be fixed. However, as each UIImage is actually displayed via a UIImageView, it will be decompressed and its memory consumption will balloon significantly, and the NSCache will still be holding onto these larger images. If you actually displayed enough UIImages that were referenced by the NSCache, then the memory spike would be the same as if forced decompression were enabled.

**Comment by bbnnt** [comment]
@foundry thx a lot for the reminder, I'll have a further look
@mallorypaine you're saying there is no real solution to all this, hence this issue (debate) will never gets closed ?

**Comment by mallorypaine** [comment]
@bbnnt there is definitely a solution to keeping unwanted memory from piling up. However, if you want to cache every image in memory then there your app's memory consumption will reflect the caching strategy. As I mentioned, I'm able to get precisely the behavior I need by disabling the NSCache. IMHO, NSCache is basically never worth using since its behavior is buggy and imprecise. It's also fairly trivial to build your own cache object that tracks cost of each item and then evicts based on LRU when the total cache cost exceeds what the app has configured. Maybe if I can find a spare hour or two I'll whip that up as it would be a true solve for this issue.

**Comment by mythodeia** [comment]
@mallorypaine so based on your pull request you turn NSCache off but keep decompression or do you disable that as well?

**Comment by bbnnt** [comment]
@mallorypaine so it might not be really an sdwebimage problem; but more or less a memory management strategy if i get it well ? I'm not expert enough; an app  with lot of images developed without this experience/knowledge would at some point always crash then ?

**Comment by holetse** [comment]
This is fundamentally an NSCache issue as I understand it. Implementing a
custom cache a la @mallorypaine would address the issue. So would Apple
fixing NSCache, but that seems to not be happening.

**Comment by rromanchuk** [comment]
I noticed SDWebImage is not using `NSCache`  countLimit, any reason? This is probably the worst documentation i've read in a while so it could be completely worthless, and i'm not even sure what the default is?  @mallorypaine  do you have any insight to this setting? 

https://developer.apple.com/library/mac/documentation/Cocoa/Reference/NSCache_Class/index.html#//apple_ref/occ/instp/NSCache/countLimit

**Comment by rromanchuk** [comment]
Actually,

`The default value is 0, which means no limit.` 

Could this be causing `- (void)setMaxMemoryCost:(NSUInteger)maxMemoryCost` to be completely ignored? I understand it's a black box, but maybe we can help coach it a bit more?

**Comment by rromanchuk** [comment]
I'm going to do a few tests with instruments with these changes, and i'll report back 

https://github.com/rs/SDWebImage/compare/master...rromanchuk:master

**Comment by trunglee** [comment]
I disable `decompression`  by changing: 
_shouldDecompressImages = YES
to: 
_shouldDecompressImages = NO 
in: 
SDImageCache.m
SDWebImageDownloader.m
SDWebImageDownloaderOperation.m

But it looks like the memory is still so high. After scrolling the table view down to some large images, the app take up huge memory.

![screen shot 2015-05-13 at 1 00 08 am](https://cloud.githubusercontent.com/assets/1575009/7594760/07797c52-f90d-11e4-9ffa-aa54cbf73ceb.png)

Can any one can help me to explain what happened here and write down some points about this issue ?

**Comment by burakkilic** [comment]
I have exactly the same issue although I made the same changes. I have some jpg images (20) in a gridview, but 80kb for each. But app crashes after one or two scroll.

**Comment by mythodeia** [comment]
@burakkilic you get a crash with just 80kb images? thats weird. i have a tableview as well with 300x250 imageview and get no crashes whatsoever.

**Comment by trunglee** [comment]
@burakkilic : In your case, its weird. My app just lagging when I'm scrolling, it still happens on iPhone 5s but get no crashes. I don't know why the memory never get down or I miss somethings here.

**Comment by burakkilic** [comment]
![screen shot 2015-05-13 at 11 04 21](https://cloud.githubusercontent.com/assets/123382/7606271/f764fa78-f95f-11e4-9f12-175ba2496a7f.png)

Here is the profiler screenshot. Images are like:

![1a067e78-de83-11e4-a6a5-7257470ec296](https://cloud.githubusercontent.com/assets/123382/7606285/1570f918-f960-11e4-9391-11067ee1eecb.jpg)

I don't know what happens

**Comment by mythodeia** [comment]
@burakkilic how large are your images in resolution?

**Comment by burakkilic** [comment]
1536*1024 and 300 Pixel per Inch

**Comment by mythodeia** [comment]
oh come on...thats huge. Thats why it crashes. Even if your size is 80kb your images are large in resolution. You should also experience laggy scroll. Generally speaking you do not add large res images as thumbs in a tableview/collection view. You add smaller ones. Do you actually need 300 ppi for thumb images?

**Comment by burakkilic** [comment]
Actually no. My client uploaded them. What is the optimum resolution?

Adil Burak Kılıç

**Comment by mythodeia** [comment]
In my apps i use 300x300 max  as thumbs while scrolling. You could use lower if you want.
And you can load the large image when you tap the thumb.

**Comment by bbnnt** [comment]
@burakkilic  "300 Pixel per Inch" is default for print, not digital

**Comment by burakkilic** [comment]
Thank you. I am downgrading to 72

Adil Burak Kılıç

**Comment by rromanchuk** [comment]
Guys, the images you are posting do not show any memory issues.  If your apps are lagging it's most likely because you are doing silly things on the main thread.  @trunglee you say huge memory growth but what exactly is your *persisted* memory after scrolling?

**Comment by trunglee** [comment]
I have a table view with a list of image urls. When the table scrolling down, it's show image on every row. That's all. For some first pages (about 20 rows per page), the table view is still smooth in an acceptable way. But from the page 3, the table view is going to lagging. I just keep a list of urls as text string for table view metadata and use SDWebImage to load image for imageView on table view cell.

**Comment by pprochazka72** [comment]
It seams like a long discussing around this particular bug, but is anyone still experiencing a huge growth in memory allocation when loading high resolution images? I experience this issue primary on iPhone 6. I tried the suggested solution (from fork https://github.com/harishkashyap/SDWebImage/tree/fix-memory-issues) to turn off 
`[[SDImageCache sharedImageCache] setShouldDecompressImages:NO];`
`[[SDWebImageDownloader sharedDownloader] setShouldDecompressImages:NO];`
but that does not solve the memory growth. For other low resolution images I see no significant growth

**Comment by bpoplauschi** [comment]
@pprochazka72 are you using static images with high res or animated ones (GIF or other format)?

**Comment by pprochazka72** [comment]
@bpoplauschi static images with high res.

**Comment by CavalcanteLeo** [comment]
+1

**Comment by weng1250** [comment]
[[SDImageCache sharedImageCache] setShouldDecompressImages:NO];
[[SDWebImageDownloader sharedDownloader] setShouldDecompressImages:NO]; is good for me to avoid huge memory cost when it is processing GIF format pics.

**Comment by bpoplauschi** [comment]
We have replaced our GIF component with FLAnimatedImage. You can try our 4.x branch ([instructions](https://github.com/rs/SDWebImage/tree/4.x#animated-images-gif-support)).

**Comment by songshp39** [comment]
image = [[UIImage alloc] initWithData:data]; 
This init method will bring in a memory spike when the image data is larger than 1M.

**Comment by bpoplauschi** [comment]
Closing. The FLAnimatedImage solution replacing our own GIF is integrated into the 4.x branch.

**Comment by bpoplauschi** [comment]
@mythodeia @rs it seems that 3 different people created PR's with the same idea of fixing some of the memory issues we have: scaling down the large images: #787  #884 #894. Do you think it is a good idea? Which one do you prefer? If we merge any of them, I would go with a `SDWebImageScaleDownLargeImage` option so people can go around it if they really need to.

**Comment by mythodeia** [comment]
@bpoplauschi #884 and #894 are almost the same.From these two i prefer #884.
Now for #787 i would like to see some benchmarks and the code needs some cleanup

**Comment by bpoplauschi** [comment]
@mythodeia I agree. Now since I saw that #787 was the most solid PR and it also included the Apple sample for downscaling, I considered including it is best. I merged it via 959d965 (of course it had conflicts and I updated the code). Then I added another commit 00bf467 cleaning stuff up in the Decoder class. Take a look and let me know what you think. We still need to bechmark it.

Everyone, #787 and 00bf467 should give you a good solution for dealing with large images. Together with the GIF handled by FLAnimatedImage, this ticket should be done now.
